Could not find or load main class on Windows

As reported at https://github.com/jreleaser/jreleaser/issues/154

Running JReleaser snapshot triggers the long classpath fix introduced in https://github.com/jbangdev/jbang/issues/161
However the launch process fails

```text
[jbang] Building jar...
Error: Could not find or load main class @C:\Users\tomer\AppData\Local\Temp\jbang2147740684520164309.args
```

Copied from the `.args` file:

```text
-classpath C:\Users\tomer\.jbang\cache\jars\jreleaser_snapshot.java.e67cca5a85c52d8c431fb6df107f8e47df46cd734c96d1b06c0ece9d6a162f46.jar;C:\Users\tomer\.m2\repository\com\github\jreleaser\jreleaser\jreleaser\main-7f55e88bad-1\jreleaser-main-7f55e88bad-1.jar;C:\Users\tomer\.m2\repository\com\github\jreleaser\jreleaser\jreleaser-workflow\main-7f55e88bad-1\jreleaser-workflow-main-7f55e88bad-1.jar;C:\Users\tomer\.m2\repository\com\github\jreleaser\jreleaser\jreleaser-engine\main-7f55e88bad-1\jreleaser-engine-main-7f55e88bad-1.jar;C:\Users\tomer\.m2\repository\com\github\jreleaser\jreleaser\git-sdk\main-7f55e88bad-1\git-sdk-main-7f55e88bad-1.jar;C:\Users\tomer\.m2\repository\org\eclipse\jgit\org.eclipse.jgit\5.11.0.202103091610-r\org.eclipse.jgit-5.11.0.202103091610-r.jar;C:\Users\tomer\.m2\repository\com\googlecode\javaewah\JavaEWAH\1.1.7\JavaEWAH-1.1.7.jar;C:\Users\tomer\.m2\repository\com\github\jreleaser\jreleaser\discord-java-sdk\main-7f55e88bad-1\discord-java-sdk-main-7f55e88bad-1.jar;C:\Users\tomer\.m2\repository\com\github\jreleaser\jreleaser\java-sdk-commons\main-7f55e88bad-1\java-sdk-commons-main-7f55e88bad-1.jar;C:\Users\tomer\.m2\repository\io\github\openfeign\form\feign-form\3.8.0\feign-form-3.8.0.jar;C:\Users\tomer\.m2\repository\commons-io\commons-io\2.8.0\commons-io-2.8.0.jar;C:\Users\tomer\.m2\repository\com\github\jreleaser\jreleaser\gitter-java-sdk\main-7f55e88bad-1\gitter-java-sdk-main-7f55e88bad-1.jar;C:\Users\tomer\.m2\repository\com\github\jreleaser\jreleaser\mail-java-sdk\main-7f55e88bad-1\mail-java-sdk-main-7f55e88bad-1.jar;C:\Users\tomer\.m2\repository\com\sun\mail\jakarta.mail\1.6.6\jakarta.mail-1.6.6.jar;C:\Users\tomer\.m2\repository\com\sun\activation\jakarta.activation\1.2.1\jakarta.activation-1.2.1.jar;C:\Users\tomer\.m2\repository\com\github\jreleaser\jreleaser\sdkman-java-sdk\main-7f55e88bad-1\sdkman-java-sdk-main-7f55e88bad-1.jar;C:\Users\tomer\.m2\repository\io\github\openfeign\feign-core\10.12\feign-core-10.12.jar;C:\Users\tomer\.m2\repository\io\github\openfeign\feign-jackson\10.12\feign-jackson-10.12.jar;C:\Users\tomer\.m2\repository\com\github\jreleaser\jreleaser\slack-java-sdk\main-7f55e88bad-1\slack-java-sdk-main-7f55e88bad-1.jar;C:\Users\tomer\.m2\repository\com\github\jreleaser\jreleaser\teams-java-sdk\main-7f55e88bad-1\teams-java-sdk-main-7f55e88bad-1.jar;C:\Users\tomer\.m2\repository\com\github\jreleaser\jreleaser\twitter-java-sdk\main-7f55e88bad-1\twitter-java-sdk-main-7f55e88bad-1.jar;C:\Users\tomer\.m2\repository\org\twitter4j\twitter4j-core\4.0.7\twitter4j-core-4.0.7.jar;C:\Users\tomer\.m2\repository\com\github\jreleaser\jreleaser\zulip-java-sdk\main-7f55e88bad-1\zulip-java-sdk-main-7f55e88bad-1.jar;C:\Users\tomer\.m2\repository\com\github\jreleaser\jreleaser\github-java-sdk\main-7f55e88bad-1\github-java-sdk-main-7f55e88bad-1.jar;C:\Users\tomer\.m2\repository\org\apache\tika\tika-core\1.26\tika-core-1.26.jar;C:\Users\tomer\.m2\repository\javax\xml\bind\jaxb-api\2.3.1\jaxb-api-2.3.1.jar;C:\Users\tomer\.m2\repository\javax\activation\javax.activation-api\1.2.0\javax.activation-api-1.2.0.jar;C:\Users\tomer\.m2\repository\javax\activation\activation\1.1.1\activation-1.1.1.jar;C:\Users\tomer\.m2\repository\org\kohsuke\github-api\1.128\github-api-1.128.jar;C:\Users\tomer\.m2\repository\org\apache\commons\commons-lang3\3.9\commons-lang3-3.9.jar;C:\Users\tomer\.m2\repository\com\github\jreleaser\jreleaser\gitlab-java-sdk\main-7f55e88bad-1\gitlab-java-sdk-main-7f55e88bad-1.jar;C:\Users\tomer\.m2\repository\com\github\jreleaser\jreleaser\gitea-java-sdk\main-7f55e88bad-1\gitea-java-sdk-main-7f55e88bad-1.jar;C:\Users\tomer\.m2\repository\io\github\openfeign\feign-httpclient\10.12\feign-httpclient-10.12.jar;C:\Users\tomer\.m2\repository\org\apache\httpcomponents\httpclient\4.5.13\httpclient-4.5.13.jar;C:\Users\tomer\.m2\repository\org\apache\httpcomponents\httpcore\4.4.13\httpcore-4.4.13.jar;C:\Users\tomer\.m2\repository\org\slf4j\jcl-over-slf4j\1.7.30\jcl-over-slf4j-1.7.30.jar;C:\Users\tomer\.m2\repository\com\github\jreleaser\jreleaser\codeberg-java-sdk\main-7f55e88bad-1\codeberg-java-sdk-main-7f55e88bad-1.jar;C:\Users\tomer\.m2\repository\com\github\jreleaser\jreleaser\jreleaser-tools\main-7f55e88bad-1\jreleaser-tools-main-7f55e88bad-1.jar;C:\Users\tomer\.m2\repository\com\github\jreleaser\jreleaser\jreleaser-templates\main-7f55e88bad-1\jreleaser-templates-main-7f55e88bad-1.jar;C:\Users\tomer\.m2\repository\com\github\spullara\mustache\java\compiler\0.9.7\compiler-0.9.7.jar;C:\Users\tomer\.m2\repository\org\zeroturnaround\zt-exec\1.12\zt-exec-1.12.jar;C:\Users\tomer\.m2\repository\com\github\jreleaser\jreleaser\jreleaser-assemblers\main-7f55e88bad-1\jreleaser-assemblers-main-7f55e88bad-1.jar;C:\Users\tomer\.m2\repository\com\github\jreleaser\jreleaser\artifactory-java-sdk\main-7f55e88bad-1\artifactory-java-sdk-main-7f55e88bad-1.jar;C:\Users\tomer\.m2\repository\com\github\jreleaser\jreleaser\jreleaser-config-yaml\main-7f55e88bad-1\jreleaser-config-yaml-main-7f55e88bad-1.jar;C:\Users\tomer\.m2\repository\com\github\jreleaser\jreleaser\jreleaser-model\main-7f55e88bad-1\jreleaser-model-main-7f55e88bad-1.jar;C:\Users\tomer\.m2\repository\com\github\jreleaser\jreleaser\jreleaser-utils\main-7f55e88bad-1\jreleaser-utils-main-7f55e88bad-1.jar;C:\Users\tomer\.m2\repository\kr\motd\maven\os-maven-plugin\1.7.0\os-maven-plugin-1.7.0.jar;C:\Users\tomer\.m2\repository\org\apache\maven\maven-plugin-api\3.5.3\maven-plugin-api-3.5.3.jar;C:\Users\tomer\.m2\repository\org\apache\maven\maven-model\3.5.3\maven-model-3.5.3.jar;C:\Users\tomer\.m2\repository\org\apache\maven\maven-artifact\3.5.3\maven-artifact-3.5.3.jar;C:\Users\tomer\.m2\repository\org\eclipse\sisu\org.eclipse.sisu.plexus\0.3.3\org.eclipse.sisu.plexus-0.3.3.jar;C:\Users\tomer\.m2\repository\javax\enterprise\cdi-api\1.0\cdi-api-1.0.jar;C:\Users\tomer\.m2\repository\javax\annotation\jsr250-api\1.0\jsr250-api-1.0.jar;C:\Users\tomer\.m2\repository\javax\inject\javax.inject\1\javax.inject-1.jar;C:\Users\tomer\.m2\repository\org\eclipse\sisu\org.eclipse.sisu.inject\0.3.3\org.eclipse.sisu.inject-0.3.3.jar;C:\Users\tomer\.m2\repository\org\codehaus\plexus\plexus-component-annotations\1.5.5\plexus-component-annotations-1.5.5.jar;C:\Users\tomer\.m2\repository\org\codehaus\plexus\plexus-classworlds\2.5.2\plexus-classworlds-2.5.2.jar;C:\Users\tomer\.m2\repository\org\codehaus\plexus\plexus-utils\3.1.0\plexus-utils-3.1.0.jar;C:\Users\tomer\.m2\repository\com\google\code\findbugs\jsr305\3.0.2\jsr305-3.0.2.jar;C:\Users\tomer\.m2\repository\org\bouncycastle\bcpg-jdk15on\1.68\bcpg-jdk15on-1.68.jar;C:\Users\tomer\.m2\repository\org\bouncycastle\bcprov-jdk15on\1.68\bcprov-jdk15on-1.68.jar;C:\Users\tomer\.m2\repository\commons-codec\commons-codec\1.15\commons-codec-1.15.jar;C:\Users\tomer\.m2\repository\com\fasterxml\jackson\core\jackson-core\2.12.3\jackson-core-2.12.3.jar;C:\Users\tomer\.m2\repository\com\fasterxml\jackson\core\jackson-databind\2.12.3\jackson-databind-2.12.3.jar;C:\Users\tomer\.m2\repository\com\fasterxml\jackson\core\jackson-annotations\2.12.3\jackson-annotations-2.12.3.jar;C:\Users\tomer\.m2\repository\com\fasterxml\jackson\dataformat\jackson-dataformat-yaml\2.12.3\jackson-dataformat-yaml-2.12.3.jar;C:\Users\tomer\.m2\repository\org\yaml\snakeyaml\1.27\snakeyaml-1.27.jar;C:\Users\tomer\.m2\repository\com\github\jreleaser\jreleaser\jreleaser-config-toml\main-7f55e88bad-1\jreleaser-config-toml-main-7f55e88bad-1.jar;C:\Users\tomer\.m2\repository\com\fasterxml\jackson\dataformat\jackson-dataformat-toml\2.12.3\jackson-dataformat-toml-2.12.3.jar;C:\Users\tomer\.m2\repository\com\github\jreleaser\jreleaser\jreleaser-config-json\main-7f55e88bad-1\jreleaser-config-json-main-7f55e88bad-1.jar;C:\Users\tomer\.m2\repository\info\picocli\picocli\4.6.1\picocli-4.6.1.jar;C:\Users\tomer\.m2\repository\org\slf4j\slf4j-api\1.7.30\slf4j-api-1.7.30.jar;C:\Users\tomer\.m2\repository\org\slf4j\slf4j-simple\1.7.30\slf4j-simple-1.7.30.jar jreleaser_snapshot release
```

The script is https://github.com/jreleaser/jbang-catalog/blob/main/jreleaser_snapshot.java

```java
//usr/bin/env jbang "$0" "$@" ; exit $?
//JAVA 8
//REPOS jitpack
//DEPS com.github.jreleaser.jreleaser:jreleaser:main-SNAPSHOT

public class jreleaser_snapshot {
    public static void main(String... args) throws Exception {
        org.jreleaser.cli.Main.main(args);
    }
}
```

> @TomerFi you can get it from the CI build referenced in the PR. But this should be the direct link: https://github.com/jbangdev/jbang/suites/2824128360/artifacts/62905418

A new error is now introduced:

```text
[jbang] jbang version 0.71.1.2
[jbang] Reading catalog from C:\Users\tomer\.jbang\implicit-catalog.json
[jbang] Retrieved file from cache https://raw.githubusercontent.com/jreleaser/jbang-catalog/HEAD/jbang-catalog.json = C:\Users\tomer\.jbang\cache\urls\6f3ac50823c8c97efd6d688f5a3526f190996f89581b8a1b88238f1fef06c1d6\jbang-catalog.json
[jbang] Reading catalog from C:\Users\tomer\.jbang\cache\urls\6f3ac50823c8c97efd6d688f5a3526f190996f89581b8a1b88238f1fef06c1d6\jbang-catalog.json
[jbang] Obtained catalog from https://github.com/jreleaser/jbang-catalog/blob/HEAD/jbang-catalog.json
[jbang] Duplicating classpath resource jbang-catalog.json
[jbang] Reading catalog from C:\Users\tomer\.jbang\cache\urls\2c97f3a787487bf73b7f4d02e109487e0e43be594ef95072827efd1500cd99ae
[jbang] Retrieved file from cache https://raw.githubusercontent.com/jreleaser/jbang-catalog/HEAD/jreleaser_snapshot.java = C:\Users\tomer\.jbang\cache\urls\8d07987c39cc02863d2aa656f3dc9caa0091d30475406778a849c219abc4a4a0\jreleaser_snapshot.java
[jbang] System Java version detected as 16
[jbang] System Java version matches requested version 16
[jbang] run: ^"C:\Program^ Files\OpenJDK\jdk-16.0.1\bin\java.exe^" @C:\Users\tomer\AppData\Local\Temp\jbang4684529230248603702.args
Error: Could not find or load main class jreleaser_snapshot
Caused by: java.lang.ClassNotFoundException: jreleaser_snapshot
```

--- a/src/main/java/dev/jbang/cli/BaseBuildCommand.java
+++ b/src/main/java/dev/jbang/cli/BaseBuildCommand.java
@@ -103,7 +103,8 @@ public abstract class BaseBuildCommand extends BaseScriptDepsCommand {
 			JarSource jarSrc = src.asJarSource();
 			if (jarSrc == null
 					|| !jarSrc.isUpToDate()
-					|| JavaUtil.javaVersion(requestedJavaVersion) < JavaUtil.javaVersion(jarSrc.getJavaVersion())) {
+					|| JavaUtil.javaVersion(requestedJavaVersion) < JavaUtil.minRequestedVersion(
+							jarSrc.getJavaVersion())) {
 				buildRequired = true;
 			} else {
 				result = ctx.importJarMetadataFor(jarSrc);

--- a/src/main/java/dev/jbang/cli/BaseBuildCommand.java
+++ b/src/main/java/dev/jbang/cli/BaseBuildCommand.java
@@ -363,6 +363,14 @@ public abstract class BaseBuildCommand extends BaseScriptDepsCommand {
 		return arg;
 	}
 
+	static String escapeArgsFileArgument(String arg) {
+		if (!shellSafeChars.matcher(arg).matches()) {
+			arg = arg.replaceAll("([\"'\\\\])", "\\\\$1");
+			arg = "\"" + arg + "\"";
+		}
+		return arg;
+	}
+
 	static String escapeWindowsArgument(String arg) {
 		if (Util.isUsingPowerShell()) {
 			if (!pwrSafeChars.matcher(arg).matches()) {

--- a/src/main/java/dev/jbang/cli/Run.java
+++ b/src/main/java/dev/jbang/cli/Run.java
@@ -8,7 +8,6 @@ import java.util.List;
 import java.util.Map;
 import java.util.Optional;
 import java.util.Stack;
-import java.util.StringJoiner;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 import java.util.stream.Collectors;
@@ -111,19 +110,24 @@ public class Run extends BaseBuildCommand {
 	String generateOSCommandLine(Source src, RunContext ctx) throws IOException {
 		List<String> fullArgs = generateCommandLineList(src, ctx);
 		String args = String.join(" ", escapeOSArguments(fullArgs));
-		// This avoids long classpath problem on Windows.
-		// @file is only available from java 9 onwards.
+		// Check if we need to use @-files on Windows
+		boolean useArgsFile = false;
 		if (args.length() > COMMAND_LINE_LENGTH_LIMIT && Util.isWindows()
-				&& JavaUtil.determineJavaVersion() >= 9 && !Util.isUsingPowerShell()) {
-			final String javaCmd = fullArgs.get(0);
-			StringJoiner joiner = new StringJoiner(" ");
-			// we must skip the first value
-			for (int i = 1; i < fullArgs.size(); ++i)
-				joiner.add(fullArgs.get(i));
-			args = joiner.toString();
+				&& !Util.isUsingPowerShell()) {
+			// @file is only available from java 9 onwards.
+			String requestedJavaVersion = javaVersion != null ? javaVersion : src.getJavaVersion();
+			int actualVersion = JavaUtil.javaVersion(requestedJavaVersion);
+			useArgsFile = actualVersion >= 9;
+		}
+		if (useArgsFile) {
+			// @-files avoid problems on Windows with very long command lines
+			final String javaCmd = escapeOSArgument(fullArgs.get(0));
 			final File argsFile = File.createTempFile("jbang", ".args");
 			try (PrintWriter pw = new PrintWriter(argsFile)) {
-				pw.println(args);
+				// write all arguments except the first to the file
+				for (int i = 1; i < fullArgs.size(); ++i) {
+					pw.println(escapeArgsFileArgument(fullArgs.get(i)));
+				}
 			}
 
 			return javaCmd + " @" + argsFile.toString();

